src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs

```diff
@@ -0,0 +1,216 @@
+
+                // Do not instrument the instrumentation helpers if they are part of the current compilation (which occurs only during testing). GetCreatePayload will fail with an infinite recursion if it is instrumented.
+                if ((object)createPayload != null && (object)flushPayload != null && (object)method != createPayload && (object)method != flushPayload)
+                {
+                    // Create the symbol for the instrumentation payload.
+                    SyntheticBoundNodeFactory factory = new SyntheticBoundNodeFactory(method, methodBody.Syntax, compilationState, diagnostics);
+                    TypeSymbol boolType = compilation.GetSpecialType(SpecialType.System_Boolean);
+                    TypeSymbol payloadElementType = boolType;
+                    ArrayTypeSymbol payloadType = ArrayTypeSymbol.CreateCSharpArray(compilation.Assembly, payloadElementType);
+                    FieldSymbol payloadField = GetPayloadField(method, methodOrdinal, payloadType, factory, compilation);
+
+                    // Synthesize the instrumentation and collect the spans of interest.
+
+                    BoundBlock newMethodBody;
+                    InstrumentationInjectionRewriter.InstrumentMethod(method, methodBody, payloadField, compilationState, diagnostics, debugDocumentProvider, out dynamicAnalysisSpans, out newMethodBody);
+
+                    // Synthesize the initialization of the instrumentation payload array. It should be done either statically or with concurrency-safe code:
+                    //
+                    // if (payloadField == null)
+                    //     Instrumentation.CreatePayload(mvid, method, ref payloadField, payloadLength);
+
+                    // ToDo: The containing module's mvid should be computed statically and stored in a static field rather than being
+                    // recomputed in each method prologue.
+                    BoundExpression mvid = factory.Property(factory.Property(factory.TypeofBeforeRewriting(method.ContainingType), "Module"), "ModuleVersionId");
+                    BoundExpression methodToken = factory.MethodToken(method);
+                    BoundStatement createPayloadCall = factory.ExpressionStatement(factory.Call(null, createPayload, mvid, methodToken, factory.Field(null, payloadField), factory.Literal(dynamicAnalysisSpans.Length)));
+
+                    BoundExpression payloadNullTest = factory.Binary(BinaryOperatorKind.ObjectEqual, boolType, factory.Field(null, payloadField), factory.Null(payloadType));
+                    BoundStatement payloadIf = factory.If(payloadNullTest, createPayloadCall);
+
+                    ImmutableArray<BoundStatement> newStatements = newMethodBody.Statements.Insert(0, payloadIf);
+                    newMethodBody = newMethodBody.Update(newMethodBody.Locals, newMethodBody.LocalFunctions, newStatements);
+
+                    if (IsTestMethod(method))
```

Comment:
General comment on the instrumentation design: Might this method of flushing payloads be brittle?
- Do we expect there to be a reliable and general way of detecting test methods? 
- Can you ensure the test code will be re-built when needed?